Add a Profinet DCP device discovery script

agalauner-r7 opened the pull request:
This patch adds an NSE script which allows device discovery of Profinet DCP devices on a network.

Profinet is an industrial Ethernet fieldbus protocol suite. DCP is a subprotocol of Profinet which is used for device discovery and configuration. This script sends a standard DCP Identify request and parses all the responses it gets back from the devices on the network.

Sample output:
```
$ sudo ./nmap -e enp8s0 --script broadcast-pndcp-discovery
Starting Nmap 7.80SVN ( https://nmap.org ) at 2019-11-25 21:15 CET
Pre-scan script results:
| broadcast-pndcp-discovery: 
|   08:00:27:2d:c2:02 (Oracle VirtualBox virtual NIC): 
|     Interface: enp8s0
|     IP: 
|       IP Info: IP set
|       IP: 192.168.178.36
|       Netmask: 255.255.255.0
|       Gateway: 192.168.178.1
|     Device: 
|       Device manufacturer: SIMATIC-PC
|       Name of Station: plcdev
|       Vendor ID: 0x002a
|       Device ID: 0x0005
|       Device Role: 0x00 (None)
|   00:30:de:40:29:c7 (Wago Kontakttechnik Gmbh): 
|     Interface: enp8s0
|     IP: 
|       IP Info: IP set
|       IP: 192.168.1.8
|       Netmask: 255.255.255.0
|       Gateway: 192.168.1.1
|     Device: 
|       Name of Station: wago-750-375
|       Device manufacturer: WAGO-I/O-SYSTEM 750/753
|       Vendor ID: 0x011d
|       Device ID: 0x0005
|       Device Role: 0x01 (IO-Device)
|   00:01:05:2d:82:5f (Beckhoff Automation GmbH): 
|     Interface: enp8s0
|     IP: 
|       IP Info: IP set
|       IP: 192.168.1.1
|       Netmask: 255.255.255.0
|       Gateway: 192.168.1.1
|     Device: 
|       Device manufacturer: TwinCAT PNIO Controller
|       Name of Station: tc-pncontroller
|       Vendor ID: 0x0120
|       Device ID: 0x0005
|       Device Role: 0x02 (IO-Controller)
|   00:01:05:3c:94:17 (Beckhoff Automation GmbH): 
|     Interface: enp8s0
|     IP: 
|       IP Info: IP set
|       IP: 192.168.1.7
|       Netmask: 255.255.255.0
|       Gateway: 192.168.1.1
|     Device: 
|       Device manufacturer: TwinCAT Profinet I/O
|       Name of Station: cx5140
|       Vendor ID: 0x0120
|       Device ID: 0x0005
|_      Device Role: 0x01 (IO-Device)
```

agalauner-r7 commented:
Any feedback on this? Is something keeping this from being merged?
I'd be happy to clean up my stuff to get this upstream!

tsellers-r7 commented:
@bonsaiviking - Is there any testing/linting you'd like to see on this? I've kinda of avoided touching it since Andy and I work on the same team at R7.

fyodor commented:
Hi @tsellers-r7 and @agalauner-r7 .  Thanks for submitting these and for the reminder.  We'll take a look at this and the #2373  one soon.

dmiller-nmap reviewed (changes requested):
Really good script! I look forward to integrating it. Just need a few cleanup things.

agalauner-r7 commented on scripts/broadcast-pndcp-discovery.nse:
```diff
@@ -0,0 +1,449 @@
+  elseif suboption == PNDCP_SUBOPTION_DEVICE_DEV_ID then
+    local dcp_suboption_device_id_format = ">x x I2"
+    if #block >= string.packsize(dcp_suboption_device_id_format) then
+      local vendor_id, device_id = string.unpack(dcp_suboption_device_id_format, block)
```
Good catch, that was a legit bug I introduced! The correct format string should be `>x x I2 I2`

agalauner-r7 commented on scripts/broadcast-pndcp-discovery.nse:
```diff
@@ -0,0 +1,449 @@
+  -- so I'm not parsing them here as I can't test it
+  elseif suboption == PNDCP_SUBOPTION_DEVICE_ALIAS_NAME then
+  elseif suboption == PNDCP_SUBOPTION_DEVICE_DEV_INSTANCE then
+  elseif suboption == PNDCP_SUBOPTION_DEVICE_OEM_DEV_ID then
```
I just added this. And it turns out two of my devices even support suboption `7` and `8`. I am going to investigate this further and see if I can parse them as well.

agalauner-r7 replied:
I am parsing all of these options now and have an `else` branch which possibly handles unknown or out-of-spec blocks.
Suboption `7` and `8` I could test, `6` - the "Alias Name" not, but I figured it's just a simple string just like the "Station Name" or "Manufacturer Name", so I added this as well. The wireshark dissector does the same, so that should be alright.

agalauner-r7 commented on scripts/broadcast-pndcp-discovery.nse:
```diff
@@ -0,0 +1,449 @@
+    for _, iface in ipairs(ifacelist) do
+      -- Match all ethernet interfaces
+      if iface.address and iface.link == "ethernet" and
+        iface.address:match("%d+%.%d+%.%d+%.%d+") then
```
No, it doesn't need an IP address, you are right.

I am trying this right now, but I am getting an error when I call `ethernet_open`:
```
bad argument #1 to 'ethernet_open' (device is not valid ethernet interface)
```

The first interface it complains about is `enp5s0` which doesn't even have a link and is marked `DOWN`, yet `iface.up` contains `up` when I print it:
```
2: enp5s0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN group default qlen 1000
    link/ether a0:36:9f:c8:cb:c3 brd ff:ff:ff:ff:ff:ff
```
```
NSE: [broadcast-pndcp-discovery] enp5s0: up
```

Any ideas why?

p-l- replied:
The interface is actually up (`NO-CARRIER,BROADCAST,MULTICAST,UP`) but has no carrier (no peer connected to it. You may want to catch that error? I have not been able to find another interface flag that would tell you if the interface is running.

agalauner-r7 replied:
Thanks @p-l-, at least I understand the error message now. The interface is up, but there's no cable plugged in, so it sees no carrier. I am going to catch that error and then don't use this interface.

Funnily enough, right now everything works because I changed my network setup and don't have a running interface without a link right now. I guess I was "lucky" the first time to even run into this issue.

agalauner-r7 replied:
Ha! So a `try`/`catch` didn't work. I grepped through the nmap code to try and find out how `ethernet_open` works and when it throws this error so that I can try and check for it mysel before I call it.

Turns out it just calls the C-equivalent of `nmap.get_interface_info`. This is what I'm doing now as well: It returns `false` when it fails. And it appears to fail when there's no carrier.

agalauner-r7 commented on scripts/broadcast-pndcp-discovery.nse:
```diff
@@ -0,0 +1,449 @@
+
+  -- We should iterate over interfaces
+  for _, interface in pairs(interfaces) do
+    local co = stdnse.new_thread(pndcpListener, interface, timeout * 1000, responses)
```
Nice! Thank you for the explanation. I tried building it that way for a while and was totally confused on how the signalling works and how to use it. With your explanation it made more sense and I was able to implement it.
While I was at it, I renamed the existing `condvar`, which is used to carry the results from a listener thread to the main thread, to `results`. Makes more sense to me.

agalauner-r7 commented on scripts/broadcast-pndcp-discovery.nse:
```diff
@@ -0,0 +1,449 @@
+  end
+
+  option, suboption, blocklen, pos = string.unpack(dcp_block_format, message, pos)
+  local blockdata = string.sub(message, pos, pos + blocklen - 1)
```
Could you explain that a bit more? Why `s2`? That would give me a string of 2 bytes length, but at this position is the length of the following block data which I called `blocklen`. I don't know how long the payload data of a block is in advance. That's why I first parsed the header and then copied out the data and realigned again at the end to a 2-byte boundary.

I added a bounds check now and left my old code in there for now. I'd like to understand your version though.

dmiller-nmap replied:
The format string 's2' unpacks a length-prefixed string where the length is 2 bytes, like `"\x00\x05hello". Your code is fine now that the length check is there.

agalauner-r7 commented:
I just addressed some of the feedback and pushed the changes. I'll check out how to parse the so far unknown suboptions and try to integrate that.

And I think the interface detection is fishy right now. See my issue above in your remarks about this.

eercanayar commented:
Congrats @agalauner-r7 great work! Can't wait to see this on the next release. Until the release, I plan to use the latest version the script file in this PR. Is there any known bugs that I should be careful about?

agalauner-r7 commented:
No known ones at least. It should work.

I am going to try and finish them soon. Had to finish a house and move and took some time off, so this work stalled a bit.

agalauner-r7 commented:
There we go. Feedback completely addressed in https://github.com/nmap/nmap/pull/1846/commits/ca197144223f862a6c722f265b0352ed61ba1dca

dmiller-nmap commented:
I'm sorry this got shelved for so long. In the meantime, we accepted several scripts (c4a4e0db45ecd5594d131f8316efe7e18df5ab6b) from @DINA-community, including [multicast-profinet-discovery.nse](https://nmap.org/nsedoc/scripts/multicast-profinet-discovery.html), which essentially does the same thing as your script. Your script has some advantages, including the IP Info suboption and several additional device property suboptions. With your permission, we'd like to integrate those into the existing script, with a co-author credit.

agalauner-r7 commented:
Sure, be my guest!
